File: Gearmaker/Gear.m
//
//  Gear.m
//  Gearmaker
//
//  Created by Torsten Kammer on 28.01.13.
//  Copyright (c) 2013 Torsten Kammer. All rights reserved.
//

#import "Gear.h"

#import "Involute.h"
#import "NSArray+Map.h"

@interface Gear()

- (void)_generateData;

@end

@implementation Gear

- (CGFloat)kopfhoehe
{
	return self.module;
}
- (CGFloat)kopfspiel
{
	return self.module * self.kopfspielfaktor;
}
- (CGFloat)fusshoehe
{
	return self.module + self.kopfspiel;
}
- (CGFloat)teilkreisdurchmesser
{
	return ((CGFloat) self.teeth) * self.module;
}
- (CGFloat)fusskreisdurchmesser
{
	return self.teilkreisdurchmesser - 2.0 * self.fusshoehe;
}
- (CGFloat)kopfkreisdurchmesser
{
	return self.teilkreisdurchmesser + 2.0 * self.kopfhoehe;
}
- (CGFloat)grundkreisdurchmesser
{
	return self.teilkreisdurchmesser * cos(self.eingriffwinkel);
}

#pragma mark - Teeth

- (NSArray *)getToothFlankPointsAtIntervals:(CGFloat)interval;
{	
	CGFloat gebogeneHoehe = (self.kopfkreisdurchmesser - self.grundkreisdurchmesser) * 0.5;
	
	NSUInteger pointsInInterval = ceil(gebogeneHoehe / interval) + 1; // +1 to get everything in the interval
	CGFloat actualInterval = gebogeneHoehe / ceil(gebogeneHoehe / interval);
	
	NSMutableArray *points = [[NSMutableArray alloc] initWithCapacity:pointsInInterval + 1];
	
	// First point at fusskreis.
	[points addObject:[NSValue valueWithPoint:CGPointMake(self.fusskreisdurchmesser * 0.5, 0.0)]];
	
	Involute *curve = [[Involute alloc] init];
	curve.baseCircleRadius = self.grundkreisdurchmesser * 0.5;
	
	for (NSUInteger i = 0; i < pointsInInterval; i++)
	{
		CGFloat radius = self.grundkreisdurchmesser*0.5 + i*actualInterval;
		CGFloat angle = [curve angleForIntersectionWithCircleRadius:radius];
		[points addObject:[NSValue valueWithPoint:[curve pointAt:angle]]];
	}
	
	return [points copy];
}

- (NSArray *)getOutlinePointsAtInveral:(CGFloat)interval;
{
	NSArray *toothFlank = [self getToothFlankPointsAtIntervals:interval];
	NSMutableArray *reverseToothFlank = [NSMutableArray arrayWithCapacity:toothFlank.count];
	for (NSValue *point in toothFlank.reverseObjectEnumerator)
	{
		NSValue *value = [NSValue valueWithPoint:NSMakePoint(point.pointValue.x, -point.pointValue.y)];
		[reverseToothFlank addObject:value];
	}

	// Find out how much angle has to be added to the reverse side
	Involute *curve = [[Involute alloc] init];
	curve.baseCircleRadius = self.grundkreisdurchmesser * 0.5;
	CGFloat angleToTeilkreis = [curve angleForIntersectionWithCircleRadius:self.teilkreisdurchmesser * 0.5];
	CGPoint intersectionPoint = [curve pointAt:angleToTeilkreis];
	CGFloat additionalAngle = atan2(intersectionPoint.y, intersectionPoint.x);
	
	NSMutableArray *points = [[NSMutableArray alloc] initWithCapacity:toothFlank.count * 2 * self.teeth];
	
	for (NSUInteger i = 0; i < self.teeth; i++)
	{
		NSAffineTransform *rotate = [NSAffineTransform transform];
		[rotate rotateByRadians:M_PI * 2.0 * (CGFloat) i / (CGFloat) self.teeth];
		
		NSArray *rotatedToothFlank = [toothFlank map:^(NSValue *point){
			return [NSValue valueWithPoint:[rotate transformPoint:point.pointValue]];
		}];
		
		[points addObjectsFromArray:rotatedToothFlank];
		
		NSAffineTransform *reverseFlankRotate = [NSAffineTransform transform];
		[reverseFlankRotate rotateByRadians:2.0 * additionalAngle + M_PI * 2.0 * (0.5 + (CGFloat) i) / (CGFloat) self.teeth];
		
		NSArray *rotatedReverseToothFlank = [reverseToothFlank map:^(NSValue *point){
			return [NSValue valueWithPoint:[reverseFlankRotate transformPoint:point.pointValue]];
		}];
		
		[points addObjectsFromArray:rotatedReverseToothFlank];
	}
	
	return [points copy];
}

#pragma mark - 3D

@synthesize vertices=_vertices;
@synthesize polygons=_polygons;

- (NSArray *)vertices
{
    if (!_vertices)
        [self _generateData];
    return _vertices;
}

- (NSArray *)polygons
{
    if (!_polygons)
        [self _generateData];
    
    return _polygons;
}

- (NSArray *)triangulatedElements
{
    NSMutableArray *elements = [NSMutableArray array];
    
    for (NSArray *polygon in self.polygons)
    {
        for (NSUInteger i = 2; i < polygon.count; ++i)
        {
            [elements addObject:polygon[0]];
            [elements addObject:polygon[i-1]];
            [elements addObject:polygon[i]];
        }
    }
    
    return elements;
}

- (NSString *)objString
{
    NSMutableString *string = [NSMutableString stringWithString:@"# Created with Gearmaker\n"];
    
    for (NSValue *value in self.vertices)
    {
        float point[3];
        [value getValue:point];
        
        [string appendFormat:@"v %.6f %.6f %.6f\n", point[0], point[1], point[2]];
    }
    
    for (NSArray *polygon in self.polygons)
    {
        [string appendString:@"f"];
        for (NSNumber *number in polygon)
            [string appendFormat:@" %@", number];
        
        [string appendString:@"\n"];
    }
    
    return [string copy];
}

- (NSString *)triangulatedObjString
{
    NSMutableString *string = [NSMutableString stringWithString:@"# Created with Gearmaker\n"];
    
    for (NSValue *value in self.vertices)
    {
        float point[3];
        [value getValue:point];
        
        [string appendFormat:@"v %.6f %.6f %.6f\n", point[0], point[1], point[2]];
    }
    
    NSArray *triangulatedElements = self.triangulatedElements;
    
    for (NSUInteger i = 0; i < triangulatedElements.count; i += 3)
    {
        [string appendFormat:@"f %@ %@ %@\n",
         triangulatedElements[i+0],
         triangulatedElements[i+1],
         triangulatedElements[i+2]];
    }
    
    return [string copy];
}

- (void)_generateData
{
    NSArray *outline = [self getOutlinePointsAtInveral:self.pointInterval];
    
    NSMutableArray *mutableVertices = [NSMutableArray array];
    
    CGFloat halfWidth = self.thickness * 0.5;
    
    // Print vertices on top
    for (NSValue *value in outline.reverseObjectEnumerator)
    {
        NSPoint point = value.pointValue;
        [mutableVertices addObject:[NSValue valueWithBytes:(float [3]) { point.x, halfWidth, point.y } objCType:@encode(float [3])]];
    }
    

    
    // Print vertices below
    for (NSValue *value in outline.reverseObjectEnumerator)
    {
        NSPoint point = value.pointValue;
        [mutableVertices addObject:[NSValue valueWithBytes:(float [3]) { point.x, -halfWidth, point.y } objCType:@encode(float [3])]];
    }
    
    
    // Create center vertex on top
    [mutableVertices addObject:[NSValue valueWithBytes:(float [3]) { 0.0f, halfWidth, 0.0f } objCType:@encode(float [3])]];
    // and below
    [mutableVertices addObject:[NSValue valueWithBytes:(float [3]) { 0.0f, -halfWidth, 0.0f } objCType:@encode(float [3])]];
    
    _vertices = [mutableVertices copy];
    
    NSMutableArray *mutablePolygons = [NSMutableArray array];
    
    // Add wall faces
    for (NSUInteger i = 0; i < outline.count; i++)
        [mutablePolygons addObject:@[
         @( (i%outline.count)+1 ),
         @( ((i)%outline.count)+outline.count+1 ),
         @( ((i+1)%outline.count)+outline.count+1 ),
         @( ((i+1)%outline.count)+1 )
         ]];
    
    // Add tooth faces
    NSUInteger pointsPerTeeth = outline.count / self.teeth;
    for (NSUInteger i = 0; i < self.teeth; i++)
    {
        // Above
        NSMutableArray *above = [NSMutableArray array];
        for (NSUInteger j = 1; j <= pointsPerTeeth; j++)
            [above addObject:@( j + i*pointsPerTeeth )];
        [mutablePolygons addObject:above];
        
        // Below - reverse for correct face winding
        NSMutableArray *below = [NSMutableArray array];
        for (NSUInteger j = pointsPerTeeth; j > 0; j--)
            [below addObject:@( j + i*pointsPerTeeth + outline.count )];
        [mutablePolygons addObject:below];
    }
    
    // Add central faces - top
    for (NSUInteger i = 0; i < self.teeth; i++)
    {
        NSUInteger triangleStart = i*pointsPerTeeth+1;
        NSUInteger triangleEnd = (i+1) * pointsPerTeeth;
        NSUInteger nextStart = ((i+1)*pointsPerTeeth+1) % outline.count;
        
        [mutablePolygons addObject:@[
                                     @( triangleStart ),
                                     @( triangleEnd ),
                                     @( _vertices.count-1 )
                                     ]];
        [mutablePolygons addObject:@[
                                     @( triangleEnd ),
                                     @( nextStart ),
                                     @( _vertices.count-1 )
                                     ]];
    }
    
    // Central faces - bottom
    for (NSUInteger i = self.teeth; i > 0; i--)
    {
        NSUInteger triangleStart = i * pointsPerTeeth;
        NSUInteger triangleEnd = triangleStart - pointsPerTeeth + 1;
        NSUInteger nextStart = triangleEnd - 1;
        if (i == 1) {
            nextStart = outline.count;
        }
        
        [mutablePolygons addObject:@[
                                     @( triangleStart + outline.count ),
                                     @( triangleEnd + outline.count ),
                                     @( _vertices.count )
                                     ]];
        [mutablePolygons addObject:@[
                                     @( triangleEnd + outline.count ),
                                     @( nextStart + outline.count ),
                                     @( _vertices.count )
                                     ]];
    }
    
    _polygons = [mutablePolygons copy];
}

#pragma mark - Loading

- (BOOL)loadFrom:(NSData *)data error:(NSError *__autoreleasing*)outError; {
    NSDictionary *dict = [NSPropertyListSerialization propertyListWithData:data options:NSPropertyListImmutable format:NULL error:outError];
    if (!dict) return NO;
    
    if (!dict[@"zaehne"] || !dict[@"modul"] || !dict[@"kopfspielfaktor"] || !dict[@"eingriffwinkel"] || !dict[@"dicke"] || !dict[@"pointInterval"])
    {
        if (outError)
            *outError = [NSError errorWithDomain:@"GearDocument" code:1 userInfo:@{
                      NSLocalizedDescriptionKey : NSLocalizedString(@"Dem Dokument fehlen wichtige Daten.", @"not all keys present in file - description"),
                      NSLocalizedRecoverySuggestionErrorKey : NSLocalizedString(@"Das Dokument ist beschädigt oder in einem Format, dass Gearmaker nicht erkennt.", @"not all keys present in file - recovery suggestion"),
                         }];
        return NO;
    }
    
    self.module = [dict[@"modul"] doubleValue];
    self.teeth = [dict[@"zaehne"] unsignedIntegerValue];
    self.eingriffwinkel = [dict[@"eingriffwinkel"] doubleValue];
    self.kopfspielfaktor = [dict[@"kopfspielfaktor"] doubleValue];
    self.pointInterval = [dict[@"pointInterval"] doubleValue];
    
    self.thickness = [dict[@"dicke"] doubleValue];
    return YES;
}

@end

File: Gearmaker/Gear.swift
//
//  Gear.swift
//  Gearmaker
//
//  Created by Torsten Kammer on 12.11.22.
//  Copyright © 2022 Torsten Kammer. All rights reserved.
//

import Foundation
import simd

@objc class Gear: NSObject, Codable {
    enum CodingKeys: String, CodingKey {
        case module = "modul"
        case teeth = "zaehne"
        case eingriffwinkel
        case kopfspielfaktor
        case pointInterval
        case thickness = "dicke"
    }
    
    @objc static func load(fromData data: Data) throws -> Gear {
        let decoder = PropertyListDecoder()
        return try decoder.decode(Gear.self, from: data)
    }
    
    @objc static func load(fromUrl url: URL) throws -> Gear {
        let data = try Data(contentsOf: url)
        return try load(fromData: data)
    }
    
    // Basic data
    
    @objc var module: Double = 0.0
    @objc var teeth: Int = 0
    @objc var kopfspielfaktor: Double = 0.0
    @objc var eingriffwinkel: Double = 0.0
    
    // Data for graphics
    @objc var pointInterval: Double = 0
    
    // Data for 3D
    @objc var thickness: Double = 0
    
    @objc var kopfhoehe: Double {
        return self.module
    }
    @objc var kopfspiel: Double {
        return module * kopfspielfaktor
    }
    @objc var fusshoehe: Double {
        return module + kopfspiel
    }
    @objc var teilkreisdurchmesser: Double {
        return Double(teeth) * module
    }
    @objc var fusskreisdurchmesser: Double {
        return teilkreisdurchmesser - 2 * fusshoehe
    }
    @objc var kopfkreisdurchmesser: Double {
        return teilkreisdurchmesser + 2 * kopfhoehe
    }
    @objc var grundkreisdurchmesser: Double {
        return teilkreisdurchmesser * cos(eingriffwinkel)
    }
    
    func toothFlankPoints(interval: Double) -> [SIMD2<Double>] {
        let gebogeneHoehe = (kopfkreisdurchmesser - grundkreisdurchmesser) * 0.5
        
        let pointsInInterval = Int(ceil(gebogeneHoehe / interval)) + 1 // +1 to get everything in the interval
        let actualInterval = gebogeneHoehe / ceil(gebogeneHoehe / interval)
        
        // First point at Fusskreis (outside the actual curve)
        var points: [SIMD2<Double>] = []
        points.reserveCapacity(pointsInInterval + 1)
        points.append(SIMD2(x: fusskreisdurchmesser * 0.5, y: 0))
        
        let curve = Involute(baseCircleRadius: grundkreisdurchmesser * 0.5)
        
        for i in 0 ..< pointsInInterval {
            let radius = grundkreisdurchmesser * 0.5 + Double(i)*actualInterval
            let angle = curve.angleForIntersectionWithCircle(radius: radius)
            points.append(curve.point(at: angle))
        }
        return points
    }
    
    private static func rotationMatrix(angle: Double) -> simd_double2x2 {
        return simd_double2x2(rows: [
            SIMD2<Double>(x: cos(angle), y: -sin(angle)),
            SIMD2<Double>(x: sin(angle), y: cos(angle)) ])
    }
    
    func outlinePoints(interval: Double) -> [SIMD2<Double>] {
        let toothFlank = toothFlankPoints(interval: interval)
        let reverseToothFlank = toothFlank.lazy.reversed().map {
            SIMD2<Double>(x: $0.x, y: -$0.y)
        }
        
        // Find out how much angle has to be added to the reverse side
        let curve = Involute(baseCircleRadius: grundkreisdurchmesser * 0.5)
        let angleToTeilkreis = curve.angleForIntersectionWithCircle(radius: teilkreisdurchmesser * 0.5)
        let intersectionPoint = curve.point(at: angleToTeilkreis)
        let additionalAngle = atan2(intersectionPoint.y, intersectionPoint.x)
        
        var points: [SIMD2<Double>] = []
        points.reserveCapacity(toothFlank.count * 2 * teeth)
        
        for tooth in 0 ..< teeth {
            
            let rotate = Gear.rotationMatrix(angle: (Double(tooth) / Double(teeth)) * 2 * Double.pi)
            for point in toothFlank {
                points.append(rotate * point)
            }
            
            let rotateReverseFlank = Gear.rotationMatrix(angle: ((Double(tooth) + 0.5) / Double(teeth)) * 2 * Double.pi + 2*additionalAngle)
            for point in reverseToothFlank {
                points.append(rotateReverseFlank * point)
            }
        }
        return points
    }
    
    struct ModelData {
        let points: [SIMD3<Double>]
        let polygons: [[Int]]
        
        var triangulated: ModelData {
            var newPolygons: [[Int]] = []
            for polygon in polygons {
                for i in 2..<polygon.count {
                    newPolygons.append( [ polygon[0], polygon[i-1], polygon[i-2] ])
                }
            }
            
            return ModelData(points: points, polygons: newPolygons)
        }
    }
    
    var modelData: ModelData {
        let outline = outlinePoints(interval: pointInterval)
        
        let halfWidth = thickness * 0.5
        var vertices: [SIMD3<Double>] = []
        vertices.reserveCapacity(outline.count*2 + 2)
        
        // Add vertices
        for point in outline.reversed() {
            vertices.append(SIMD3<Double>(x: point.x, y: halfWidth, z: point.y))
            vertices.append(SIMD3<Double>(x: point.x, y: -halfWidth, z: point.y))
        }
        // Add center vertices
        let centralTop = vertices.count
        vertices.append(SIMD3<Double>(x: 0, y: halfWidth, z: 0))
        let centralBottom = vertices.count
        vertices.append(SIMD3<Double>(x: 0, y: -halfWidth, z: 0))
        
        var polygons: [[Int]] = []
        
        // Add wall faces
        for i in 0..<outline.count {
            polygons.append([
                i*2,
                (i*2 + 1) % (outline.count*2),
                ((i+outline.count)*2 + 1) % (outline.count*2),
                (i+outline.count)*2,
            ])
        }
        
        // Add tooth faces
        let pointsPerTooth = outline.count / teeth
        for i in 0..<teeth {
            // Above
            polygons.append( (0 ..< pointsPerTooth).map {
                (i * pointsPerTooth + $0) * 2
            })
            // Below - reverse order
            polygons.append( stride(from: pointsPerTooth - 1, through: 0, by: -1).map {
                (i * pointsPerTooth + $0) * 2 + 1
            })
            
            // Central face above - tooth
            polygons.append([ i*pointsPerTooth*2, ((i+1) * pointsPerTooth - 1)*2, centralTop])
            // - gap to next tooth
            polygons.append([ ((i+1) * pointsPerTooth - 1)*2, ((i+1) * pointsPerTooth)*2, centralTop])
            
            // Central face below - tooth
            polygons.append([ ((i+1) * pointsPerTooth - 1)*2 + 1, i*pointsPerTooth*2 + 1,  centralTop])
            // - gap to next tooth
            polygons.append([ ((i+1) * pointsPerTooth)*2 + 1, ((i+1) * pointsPerTooth - 1)*2 + 1, centralTop])
        }
        
        return ModelData(points: vertices, polygons: polygons)
    }
}
